File: Algorithms/Easy/ShortestDistanceToACharacter.cpp
821. Shortest Distance to a Character

Given a string s and a character c that occurs in s, return an array of integers answer where answer.length == s.length and answer[i] 
is the distance from index i to the closest occurrence of character c in s.

The distance between two indices i and j is abs(i - j), where abs is the absolute value function.

class Solution {
public:
    std::vector<int> shortestToChar(std::string s, char c) 
    {
        std::vector<int> charPos{};

        for (int i = 0; i < s.size(); i++)
        {
            if (s[i] == c)
                charPos.push_back(i);
        }

        std::vector<int> answ {};
        for (int i = 0; i < s.size(); i++)
        {
            answ.push_back(helper(i, charPos));
        }
        
        return answ;
    }
    
    int helper(int i, const std::vector<int> &c)
    {
        int min = 1000000;

        for (int v : c)
        {
            if (abs(v - i) < min)
                min = abs(v - i);
        }

        return min;
    }
};

File: Algorithms/Easy/ShortestDistanceToACharacter.go
/*
821. Shortest Distance to a Character

Given a string s and a character c that occurs in s, return an array of integers answer where answer.length == s.length and answer[i] is the distance from index i to the closest occurrence of character c in s.

The distance between two indices i and j is abs(i - j), where abs is the absolute value function.
*/

func shortestToChar(s string, c byte) []int {
	res := make([]int, 0, len(s))
	m := make([]int, 0)

	for i, ch := range s {
		if ch == rune(c) {
			m = append(m, i)
		}
	}

	for i := range s {
		min := 1000000
		for _, v := range m {
			if abs(i-v) < min {
				min = abs(i - v)
			}
		}
		res = append(res, min)
	}
	return res
}

func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}
